### DISCUSSION TITLE
Enable SqlLogging by class/dao?

### DISCUSSION THREAD

**Discussion by chriskessel** [jdbi/jdbi#2412] [Q&A] [answered]
We have dozens of "Dao" classes for type specific lookups (e.g. query Users, Floobles, etc). I'd like to enable logging for one of them while I'm debugging something, but it looks like `setSqlLogger()` is universal.

Is there a way to be selective about what it logs?

**Comment by stevenschlansker** [comment]
Unfortunately I don't see a simple way to do this right now. If you have access to the `Handle`, you can call `getExtensionMethod()`, but you do not have access to that in the `SqlLogger` interface. We could add a `getExtensionMethod()` to the `StatementContext` object you are passed - then the logger could see which class and method were called. Would that solve your problem?

**Comment by chriskessel** [comment]
The context has a getExtensionMethod() on it. I'm using it to help augment my log line so I can grep for the class of interest.

Works for local debugging, but would break down a bit in prod with just too much spew. It be useful to say "log everything from the NewFeatureThingExtension", but not log all the other extensions. Not sure what that looks like, a logger per extension somehow? A per extension annotation that says whether it should get logged? 

Food for thought anyway.

**Comment by stevenschlansker** [comment, marked as answer]
Oh, you're entirely right, how did I miss that!

To implement that, can't you implement a logger that checks the extension class?

```
void logAfterExecution(StatementContext context) {
    if (context.getExtensionMethod().getType() == NewFeatureThingExtension.class) {
        .... log
    }
}
```

**Reply by chriskessel** [reply]
Yep, I can add a chunk of code to specifically look as you mentioned. Maybe I can extend that concept to look at a `sql_logging.yml` config or something so I can enable/disable outside of the code itself.

**Reply by stevenschlansker** [reply]
Yeah, we don't ship anything like this out of the box, but if you build something that is both generally applicable and small and self-contained, we'd be happy to consider it for inclusion :)

**Comment by stevenschlansker** [comment]
I'll close this as answered - right now you have to write code to accomplish this. If there's desire for a more configuration-forward way to do it, we'd be happy to consider contributions to that effect. Thanks for the question!

**Reply by chriskessel** [reply]
A followup question, hopefully you'll see this as I didn't want to open a new issue just to ask.

Is there a way to see the parameters that get used? I've tried context.rendered, raw, and parsed, but none of them show the parameter substitution.

My end goal here is to be able to grab the the actual SQL, then go run it in a DB console and tweak it until it's doing what I want, then go back and change my JDBI to reflect my new code. The problem is none of the logged SQL has the actual parameter values, so I have to log those out and copy/paste them into the SQL...and it's sort of tedious :(

**Reply by stevenschlansker** [reply]
@chriskessel , yes, you can inspect the parameters that are used by calling `statementContext.getBinding()`. It has a descriptive `toString` and (deprecated) methods for fetching bound arguments.

However, I think what you want is not really possible anyway. Jdbi is not textually substituting parameters into SQL at all - it sends along the rendered sql string `insert into table values(?)` and then passes the parameters separately, using the database driver which could even send just binary data.

I don't think there actually is a general way using JDBC to take a prepared statement and emit the SQL you would paste into a console.

**Reply by chriskessel** [reply]
Bummer. I think I got used to the JDBC postgres driver's PreparedStatement whose toString() did spit out the actual SQL. I think it's in the NativeQuery.toString() if I've followed the code correctly. That's Postgres specific obviously, though I've been using Postgres for so long now I hadn't considered where it was coming from.

I'm still using Postgres now, but I don't see a way to hook into that nice toString() via JDBI's usage.

**Reply by stevenschlansker** [reply]
Oh, I didn't realize the driver could do that.
The StatementContext also provides direct access to the PreparedStatement via `getStatement` - does that do what you want?

**Reply by chriskessel** [reply]
I was just going there :). Yes, that lets me get to the PreparedStatement! Sadly, in our environment it's one that wraps the actual Postgres one, so I'm digging into that. But that's a me issue, not a JDBI issue.

So, the good news for anyone that finds this by Google search and uses a normal library (e.g. Hikari) with Postgres, you should be able to log the actual SQL with context.getStatement().toString(). Hurray!

Thanks for your help on this, it definitely helped guide me in poking around in the guts.

**Reply by stevenschlansker** [reply]
If your wrapper implements the `java.sql.Wrapper` interface, you might be able to `statementContext.getStatement().unwrap(PgStatement.class).toString()`, although it is not guaranteed to help, depending on exactly how the wrapper works.

**Reply by chriskessel** [reply]
Sweet, that works!

**Reply by stevenschlansker** [reply]
Awesome!
